fix: Fix various bugs in quickstart scripts

JonahSussman opened the pull request:
**What this PR does / why we need it**:

This PR fixes installation failures when following the [quickstart guide](https://kserve.github.io/website/docs/getting-started/quickstart-guide) by fixing some issues along the same lines as #4813:

1. **BIN_DIR creation**: The `bin` directory is now created if it doesn't exist, preventing failures when tools like `kustomize`, `helm`, and `yq` are downloaded
2. **stderr redirection**: Warnings from `find_repo_root()` are now properly redirected to stderr, preventing them from polluting the `REPO_ROOT` variable when captured via command substitution
3. **Cleanup logic**: Added trap-based cleanup to remove `BIN_DIR` if created by the script, keeping the environment clean
4. **Early installation exit**: Removed `exit 0`s inside installation functions causing the install scripts to not finish
5. **Documentation fix**: Fixed broken contributor guide link in template

**Problem:**
Running the quickstart command:
```bash
curl -s "https://raw.githubusercontent.com/kserve/kserve/master/hack/setup/quick-install/kserve-standard-mode-full-install-with-manifests.sh" | bash
```

Results in error:
```
mv: cannot move '/tmp/tmp.Hbxk19H93N/kustomize' to '\033[93m[WARNING]\033[0m Could not find git repository root, using current directory:
/home/jonah/Projects/github.com/kserve\n/home/jonah/Projects/github.com/kserve/bin/kustomize': No such file or directory
```

Additionally, after fixing these issues, the installation functions exit early, failing to install all the required components. 

Root causes:
1. Warning messages were sent to stdout instead of stderr, causing them to be captured in BIN_DIR variable
2. The bin/ directory was assumed to exist but was never created
3. The infra installation scripts had extra `exit 0`s inside them
4. The calls for the `uninstall` and `install`s do not get renamed.

**Which issue(s) this PR fixes:**
Fixes #4831
Fixes #4843

**Type of changes**

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing:**

Validated the fix by:

- Ran hack/setup/scripts/validate-install-scripts.py - all checks pass
- Tested installation in local environment - bin directory created and cleaned up properly
- Verified warning messages appear on stderr without polluting variables
- Confirmed tools (helm, kustomize, yq) install successfully to BIN_DIR

Test configuration:
- Environment: Fedora Linux 42, Bash 5.2.37
- Installation method: Direct script execution and curl piping
- Verified both with and without pre-existing bin/ directory

**Special notes for your reviewer:**

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes. (N/A - This PR only modifies installation script logic, no image versions changed)
2. Changes are in the template file (templates/generated-script.template) so all generated scripts will benefit from these fixes
5. The cleanup logic uses a flag (`BIN_DIR_CREATED_BY_SCRIPT`) to only remove directories created by the script, preserving any pre-existing bin/ directories
6. Also fixed the broken contributor guide link

**Checklist:**

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
  - Validated using existing test suite (validate-install-scripts.py)
- [x] Has code been commented, particularly in hard-to-understand areas?
  - Added comments explaining BIN_DIR tracking and cleanup logic
- [x] Have you made corresponding changes to the documentation?
  - No documentation changes needed (fixed existing broken link)

**Release note:**

```release-note
Fixed quickstart installation script to properly create bin directory and redirect warnings to stderr
```

JonahSussman committed:
fix: add BIN_DIR cleanup and fix find_repo_root output in install scripts

Signed-off-by: JonahSussman <sussmanjonah@gmail.com>

JonahSussman committed:
Fixed broken link to developer guide

Signed-off-by: JonahSussman <sussmanjonah@gmail.com>

Jooho commented:
/hold

Jooho commented on hack/setup/common.sh:
You should add the new function in here.  Then you can execute `make generate-quick-install-scripts`. It will add the function into each quick install script.

Jooho commented on hack/setup/scripts/install-script-generator/templates/generated-script.template:
It does not need to add here

JonahSussman committed:
Moved functions to common.sh

Signed-off-by: JonahSussman <sussmanjonah@gmail.com>

JonahSussman commented:
Hey @Jooho , let me know if the latest revision implements what you were talking about

Jooho commented on hack/setup/common.sh:
```diff
@@ -75,6 +75,22 @@ detect_arch() {
+}
+
+# Set up trap to run cleanup on exit
+trap cleanup EXIT
```
Should we clean up the directory every time? If the first attempt failed, it will download the files again.

Jooho commented on hack/setup/quick-install/kserve-standard-mode-full-install-helm.sh:
```diff
@@ -49,7 +49,7 @@ find_repo_root() {
     # Git repository not found
     if [[ "$skip" == "true" ]]; then
-        log_warning "Could not find git repository root, using current directory: $PWD"
+        log_warning "Could not find git repository root, using current directory: $PWD" >&2
```
Please add `>&2` to [these methods](https://github.com/kserve/kserve/blob/master/hack/setup/common.sh#L93-L107)

Jooho reviewed (commented):
The root cause was that the correct value wasn’t being passed into the path, and it is resolved by using >&2. However, I’m not sure whether we also need to remove the existing bin folder every time. Could you clarify why that step is necessary?

JonahSussman committed:
Cleaned up error piping

Signed-off-by: JonahSussman <sussmanjonah@gmail.com>

JonahSussman commented:
Thanks for the review, @Jooho and noted, I also cleaned up the `>&2` redirection.

To your question: we don’t remove an existing bin directory. The cleanup only runs when:

1. `bin/` did not exist before the script ran, and
2. The script created it specifically for installing the tools.

That’s what the `BIN_DIR_CREATED_BY_SCRIPT` flag tracks. If a user already has a `bin/` directory, we leave it completely untouched.

The main goal was just to avoid leaving stray bin/ folders behind when users run `curl | bash` from arbitrary directories. 

An alternative would be to install into a directory under `/tmp` instead, then we wouldn’t need cleanup logic since the OS eventually clears temp dirs. Happy to switch to that approach if you’d prefer.

Let me know what you think

JonahSussman committed:
Fixed early exiting during infra installation

Signed-off-by: JonahSussman <sussmanjonah@gmail.com>

JonahSussman committed:
Fixed install and uninstall renaming

Signed-off-by: JonahSussman <sussmanjonah@gmail.com>

JonahSussman commented:
Apologies for the re-review @sivanantha321 and @spolti , I modified some of the Python files to fix the script generation as well.

Jooho commented:
@JonahSussman thanks for the reply.

KServe doesn’t have a bin directory by default, so it always gets created whenever the scripts run. It’s also included in .gitignore, so it never gets committed. Because of that, my view is that there’s no real need to delete the bin folder at all — it will always be created anyway.

For the quick-install script specifically, I’m fine with creating a temporary bin directory under /tmp. But when the repository is cloned, I think it makes more sense to place the binaries under kserve/bin.

If we want this behavior, it seems like we only need to update the quick-install template file. What do you think?

JonahSussman committed:
Used tmp dir for binaries

Signed-off-by: JonahSussman <sussmanjonah@gmail.com>

Jooho commented:
Thanks for the update. I triggered the test ci.

JonahSussman committed:
Added .gitkeep inside bin directory

Signed-off-by: JonahSussman <sussmanjonah@gmail.com>

JonahSussman commented:
@Jooho Thanks for the feedback! I think you're right - this makes more sense. I've made some changes that should address your concerns.

I added a bin/.gitkeep file so the bin/ directory will now be committed to the repo (but stay empty). I also updated .gitignore to ignore everything in
bin/ except the .gitkeep:

I also updated the logic in both common.sh and the generated-script template to check if the bin/ directory exists - if it does, it uses it; otherwise it
falls back to a temp directory:

```sh
# Set up BIN_DIR - use repo bin if it exists, otherwise use temp directory
if [[ -d "${REPO_ROOT}/bin" ]]; then
    export BIN_DIR="${REPO_ROOT}/bin"
else
    export BIN_DIR="$(mktemp -d)"
    log_info "Using temp BIN_DIR: ${BIN_DIR}"
fi
```

Now when someone clones the repo, bin/ will already exist and binaries will go there. For the quick-install curl | bash scenario where there's no repo,
it'll use /tmp instead. Since we're using temp directories for that case, there's no cleanup needed.

Does this address what you were thinking?

Jooho commented:
/retest

Jooho commented:
please do `make precommit`

JonahSussman committed:
Did make precommit

Signed-off-by: JonahSussman <sussmanjonah@gmail.com>

JonahSussman commented:
@Jooho My apologies, I re-ran `make precommit`

Jooho reviewed (approved):
/lgtm
/approve

Jooho commented:
/unhold

sivanantha321 commented on hack/setup/quick-install/kserve-knative-mode-dependency-install.sh:
```diff
@@ -83,11 +83,27 @@ detect_arch() {
     echo "$arch"
 }
 
+cleanup_bin_dir() {
```
is this still necessary ?
